### PR TITLE
Add HDInsight secure hadoop cluster creation support in powershell

### PR DESCRIPTION
**Pull request by mumou**
## Description

This change set includes changes to create an HDInsight secure cluster. 
We added a parameter SecurityProfile in New-AzureRmHDInsightCluster cmdlet and also created a new cmdlet Add-AzureRmHDInsightSecurityProfile to populate SecurityProfile parameter
If SecurityProfile is defined, it will be used to create a secure cluster

---

This checklist is used to make sure that common guidelines for a pull request are followed. You can find a more complete discussion of PowerShell cmdlet best practices [here](https://msdn.microsoft.com/en-us/library/dd878270%28v=vs.85%29.aspx).
- [x] **I have read the [contribution guidelines](https://github.com/Azure/azure-powershell/blob/dev/CONTRIBUTING.md).**
- [x] **If changes were made to any cmdlet, the XML help was regenerated using the [platyPSHelp module](https://github.com/Azure/azure-powershell/blob/dev/documentation/platyps-help.md).**
- [x] **If any large changes are made to a service, they are reflected in the respective [change log](https://github.com/Azure/azure-powershell/blob/dev/CONTRIBUTING.md#updating-the-change-log).**
### [General Guidelines](https://github.com/Azure/azure-powershell/blob/dev/CONTRIBUTING.md#general-guidelines)
- [x] Title of the pull request is clear and informative.
- [x] There are a small number of commits, each of which have an informative message. This means that previously merged commits do not appear in the history of the PR. For more information on cleaning up the commits in your PR, [see this page](https://github.com/Azure/azure-powershell/blob/dev/documentation/cleaning-up-commits.md).
- [x] The pull request does not introduce [breaking changes](https://github.com/Azure/azure-powershell/blob/dev/documentation/breaking-changes.md) (unless a major version change occurs in the assembly and module).
### [Testing Guidelines](https://github.com/Azure/azure-powershell/blob/dev/CONTRIBUTING.md#testing-guidelines)
- [x] Pull request includes test coverage for the included changes.
- [x] PowerShell scripts used in tests should do any necessary setup as part of the test or suite setup, and should not use hard-coded values for locations or existing resources.
### [Cmdlet Signature Guidelines](https://github.com/Azure/azure-powershell/blob/dev/CONTRIBUTING.md#cmdlet-signature-guidelines)
- [x] New cmdlets that make changes or have side effects should implement `ShouldProcess` and have `SupportShouldProcess=true` specified in the cmdlet attribute. You can find more information on `ShouldProcess` [here](https://gist.github.com/markcowl/338e16fe5c8bbf195aff9f8af0db585d#what-is-the-change).
- [x] Cmdlet specifies `OutputType` attribute if any output is produced - if the cmdlet produces no output, it should implement a `PassThrough` parameter.
### [Cmdlet Parameter Guidelines](https://github.com/Azure/azure-powershell/blob/dev/CONTRIBUTING.md#cmdlet-parameter-guidelines)
- [x] Parameter types should not expose types from the management library - complex parameter types should be defined in the module.
- [x] Complex parameter types are discouraged - a parameter type should be simple types as often as possible. If complex types are used, they should be shallow and easily creatable from a constructor or another cmdlet.
- [x] Cmdlet parameter sets should be mutually exclusive - each parameter set must have at least one mandatory parameter not in other parameter sets.

### CODE REVIEW

**Comment by azurecla** [comment]
Hi __@mumou__, I'm your friendly neighborhood Azure Pull Request Bot (You can call me AZPRBOT). Thanks for your contribution!
    <p>
        It looks like you're working at Microsoft (mumou). If you're full-time, we DON'T require a contribution license agreement.
    </p>
    <p>
        If you are a vendor, DO please sign the electronic contribution license agreement. It will take 2 minutes and there's no faxing! https://cla.azure.com.
    </p>

TTYL, AZPRBOT;

File: src/ResourceManager/HDInsight/Commands.HDInsight/ManagementCommands/AddAzureHDInsightSecurityProfile.cs [hunk 1]
```diff
@@ -0,0 +1,98 @@
...
+            Mandatory = true,
+            HelpMessage = "The list of urls of the LDAPS server.")]
+        public string[] LdapsUrls
+        {
+            get { return _securityProfile.LdapsUrls.ToArray(); }
+            set { _securityProfile.LdapsUrls = value; }
+        }
+
+        [Parameter(Position = 5,
+            HelpMessage = "The list of DNs of the user group which should have access to use the cluster.")]
```
Comment (shefaliv) [inline src/ResourceManager/HDInsight/Commands.HDInsight/ManagementCommands/AddAzureHDInsightSecurityProfile.cs]: distinguished names

Comment (shefaliv) [inline src/ResourceManager/HDInsight/Commands.HDInsight/ManagementCommands/AddAzureHDInsightSecurityProfile.cs]: user groups (plural)

File: src/ResourceManager/HDInsight/Commands.HDInsight/ManagementCommands/AddAzureHDInsightSecurityProfile.cs [hunk 2]
```diff
@@ -0,0 +1,98 @@
...
+            HelpMessage = "The OrganizationalUnit DN.")]
+        public string OrganizationalUnitDN
+        {
+            get { return _securityProfile.OrganizationalUnitDN; }
+            set { _securityProfile.OrganizationalUnitDN = value; }
+        }
+
+        [Parameter(Position = 4,
+            Mandatory = true,
+            HelpMessage = "The list of urls of the LDAPS server.")]
```
Comment (shefaliv) [inline src/ResourceManager/HDInsight/Commands.HDInsight/ManagementCommands/AddAzureHDInsightSecurityProfile.cs]: URLs (caps)

File: src/ResourceManager/HDInsight/Commands.HDInsight/ManagementCommands/AddAzureHDInsightSecurityProfile.cs [hunk 3]
```diff
@@ -0,0 +1,98 @@
...
+            HelpMessage = "Credential of the domain user. Username should be in user principal format. Eg. sampleuser@contoso.com")]
+        public PSCredential DomainUserCredential 
+        {
+            get { return _securityProfile.DomainUserCredential; }
+            set { _securityProfile.DomainUserCredential = value; } 
+        }
+
+        [Parameter(Position = 3,
+            Mandatory = true,
+            HelpMessage = "The OrganizationalUnit DN.")]
```
Comment (shefaliv) [inline src/ResourceManager/HDInsight/Commands.HDInsight/ManagementCommands/AddAzureHDInsightSecurityProfile.cs]: distinguished name

File: src/ResourceManager/HDInsight/Commands.HDInsight/ManagementCommands/AddAzureHDInsightSecurityProfile.cs [hunk 4]
```diff
@@ -0,0 +1,98 @@
...
+            HelpMessage = "The Domain of the secure cluster.")]
+        public string Domain
+        {
+            get { return _securityProfile.Domain; }
+            set { _securityProfile.Domain = value; } 
+        }
+
+        [Parameter(Position = 2,
+            Mandatory = true,
+            HelpMessage = "Credential of the domain user. Username should be in user principal format. Eg. sampleuser@contoso.com")]
```
Comment (shefaliv) [inline src/ResourceManager/HDInsight/Commands.HDInsight/ManagementCommands/AddAzureHDInsightSecurityProfile.cs]: "Credential of the domain user. Username should be in user principal name format, e.g. sampleuser@contoso.com"

File: src/ResourceManager/HDInsight/Commands.HDInsight/ManagementCommands/AddAzureHDInsightSecurityProfile.cs [hunk 5]
```diff
@@ -0,0 +1,98 @@
...
+
+        [Parameter(Position = 0,
+        Mandatory = true,
+        ValueFromPipeline = true,
+        HelpMessage = "The HDInsight cluster configuration to use when creating the new cluster.")]
+        public AzureHDInsightConfig Config { get; set; }
+
+        [Parameter(Position = 1,
+            Mandatory = true,
+            HelpMessage = "The Domain of the secure cluster.")]
```
Comment (shefaliv) [inline src/ResourceManager/HDInsight/Commands.HDInsight/ManagementCommands/AddAzureHDInsightSecurityProfile.cs]: domain (no caps)

File: src/ResourceManager/HDInsight/Commands.HDInsight/ManagementCommands/NewAzureHDInsightClusterCommand.cs [hunk 6]
```diff
@@ -301,6 +303,9 @@ public DateTime RdpAccessExpiry
         [Parameter(HelpMessage = "Gets or sets the Service Principal AAD Tenant Id for accessing Azure Data Lake.")]
         public Guid AadTenantId { get; set; }
 
+        [Parameter(HelpMessage = "Gets or sets Security Profile which is used for creating secure cluster.")]
```
Comment (shefaliv) [inline src/ResourceManager/HDInsight/Commands.HDInsight/ManagementCommands/NewAzureHDInsightClusterCommand.cs]: "Gets or sets Security Profile which is used for creating a secure cluster."

File: src/ResourceManager/HDInsight/Commands.HDInsight/Models/Management/AzureHDInsightSecurityProfile.cs [hunk 7]
```diff
@@ -0,0 +1,68 @@
...
+using System.Security;
+using Microsoft.Azure.Commands.HDInsight.Models.Management;
+
+namespace Microsoft.Azure.Commands.HDInsight.Models
+{
+
+    /// <summary>
+    /// Represents and AzureHDInsightSecurityProfile which contans the parameters to create secure cluster.
+    /// </summary>
+    public class AzureHDInsightSecurityProfile
```
Comment (shefaliv) [inline src/ResourceManager/HDInsight/Commands.HDInsight/Models/Management/AzureHDInsightSecurityProfile.cs]: correct doc here too based on comments in the other cmdlet

**Comment by shefaliv** [comment]
Approved.

File: src/ResourceManager/HDInsight/Commands.HDInsight/ManagementCommands/AddAzureHDInsightSecurityProfile.cs [hunk 8]
```diff
@@ -0,0 +1,98 @@
...
+        [Parameter(Position = 4,
+            Mandatory = true,
+            HelpMessage = "Urls of one or multiple LDAPS servers for the Active Directory.")]
+        public string[] LdapsUrls
+        {
+            get { return _securityProfile.LdapsUrls.ToArray(); }
+            set { _securityProfile.LdapsUrls = value; }
+        }
+
+        [Parameter(Position = 5,
```
Comment (markcowl) [inline src/ResourceManager/HDInsight/Commands.HDInsight/ManagementCommands/AddAzureHDInsightSecurityProfile.cs]: 5 == too many positional parameters

File: src/ResourceManager/HDInsight/Commands.HDInsight/ManagementCommands/AddAzureHDInsightSecurityProfile.cs [hunk 9]
```diff
@@ -0,0 +1,98 @@
...
+namespace Microsoft.Azure.Commands.HDInsight
+{
+    [Cmdlet(
+        VerbsCommon.Add,
+        Constants.CommandNames.AzureHDInsightSecurityProfile),
+    OutputType(
+        typeof(AzureHDInsightSecurityProfile))]
+    public class AddAzureHDInsightSecurityProfile : HDInsightCmdletBase
+    {
+        private readonly AzureHDInsightSecurityProfile _securityProfile;
```
Comment (markcowl) [inline src/ResourceManager/HDInsight/Commands.HDInsight/ManagementCommands/AddAzureHDInsightSecurityProfile.cs]: Should support ShouldProcess

File: src/ResourceManager/HDInsight/Commands.HDInsight/ManagementCommands/AddAzureHDInsightSecurityProfile.cs [hunk 10]
```diff
@@ -0,0 +1,98 @@
...
+        #endregion
+
+        public AddAzureHDInsightSecurityProfile()
+        {
+            _securityProfile = new AzureHDInsightSecurityProfile();
+        }
+
+        public override void ExecuteCmdlet()
+        {
+            Config.SecurityProfile = _securityProfile;
```
Comment (markcowl) [inline src/ResourceManager/HDInsight/Commands.HDInsight/ManagementCommands/AddAzureHDInsightSecurityProfile.cs]: Check ShouldProcess or ConfirmAction before making changes

File: src/ResourceManager/HDInsight/Commands.HDInsight/ManagementCommands/AddAzureHDInsightSecurityProfile.cs (hunk 9 above)
Comment (mumou) [inline src/ResourceManager/HDInsight/Commands.HDInsight/ManagementCommands/AddAzureHDInsightSecurityProfile.cs]: This cmdlet doesn't actually add anything to azure and no network call is made. It is used as a helper cmdlet to add security profile configuration object to the Config object.  This config object is later used in New-AzureRmHDInsightCluster to create an HDInsight cluster.

File: src/ResourceManager/HDInsight/Commands.HDInsight/ManagementCommands/AddAzureHDInsightSecurityProfile.cs (hunk 10 above)
Comment (mumou) [inline src/ResourceManager/HDInsight/Commands.HDInsight/ManagementCommands/AddAzureHDInsightSecurityProfile.cs]: Here we are not adding anything to azure and no network call is made. It is used as a helper cmdlet to add security profile configuration object to the Config object. This config object is later used in New-AzureRmHDInsightCluster to create an HDInsight cluster. This operation is safe and should not require permission from user.

File: src/ResourceManager/HDInsight/Commands.HDInsight/ManagementCommands/AddAzureHDInsightSecurityProfile.cs (hunk 9 above)
Comment (markcowl) [inline src/ResourceManager/HDInsight/Commands.HDInsight/ManagementCommands/AddAzureHDInsightSecurityProfile.cs]: It is making changes to the passed in configuration and so need to implement ShouldProcess.

File: src/ResourceManager/HDInsight/Commands.HDInsight/ManagementCommands/AddAzureHDInsightSecurityProfile.cs (hunk 10 above)
Comment (markcowl) [inline src/ResourceManager/HDInsight/Commands.HDInsight/ManagementCommands/AddAzureHDInsightSecurityProfile.cs]: Same comment

**Comment by markcowl** [comment]
@mumou This PR also has conflicts, please update

**Comment by mumou** [comment]
Closing this PR opening another one.

**Pull request closed without merging**